docs/proposals/signal-api.md

```diff
@@ -0,0 +1,172 @@
+may be motivated both by default behaviour that signals are bringing and the custom signal handling which the
+program is doing.
+
+The more concrete use-cases are described below.
+
+## Use-Cases
+
+#### Re-reading configuration
+
+Many applications follow the convention to re-read configuration files when handling the SIGHUP signal. Ability
+to send SIGHUP signal will allow us to reconfigure pods without updating their definition or removing them. Also,
+we can consider using this API in the watch mechanism in kubelet to reconfigure a pod automatically on the ConfigMap
+change.
+
+## Analysis
+
+The system needs to be able to:
+
+1. Model correctly which signals are supported
+2. Allow to define what signals can be used for each pod (and whether they can be used at all)
+3. Provide custom names for the chosen signals in pod
+
+## Proposed design
+
+Signal API will be a part of pod API. There will be no separate API group for that.
+
+#### Schema
+
+The proposed schema is as follow.
+
+There will be a type Notification being a part of the PodSpec:
+
+```go
+type Notification struct {
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	Type string `json:"type" protobuf:"bytes,2,opt,name=type"`
+	Signal string `json:"signal" protobuf:"bytes,3,opt,name=signal"`
+}
+```
+
+The *signal* parameter will be able to take any name of POSIX signal and that will be a criterion to validate that
+field.
+
+And there will be an API subresource of `pod` called `notify`, with the options:
+- container
+- name
+
+#### Endpoints
+
+There will be PUT and POST endpoint, rooted at /api/v1/, with the parameters as listed:
+- /namespaces/{namespace}/pods/{pod}/notify
+  - container
+  - name  - name of the notification to send
+
+#### ContainerRuntime interface
+
+The following method will be defined in the ContainerRuntime interface:
+
+```go
+Kill(id, signal string) error
+```
+
+Translating the notification name to the POSIX signal name will happen before using any container runtime, on the
+kubelet API level.
+
+In case of Docker, the `kill` API resource will be called.
+
+In rkt, the only way to signal containers is to use machinectl. Implementing the good way to signal containers in rkt
+seems to be out of scope of this proposal.
+
+## Examples
+
+Let's assume that we have a pod defined like:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: signal-example
+spec:
+  containers:
+    - name: my-container
+      image: myusername/my-program-which-handles-signals
+      notifications:
+        - name: reconfigure
+          type: signal
+          signal: SIGHUP
+        - name: stop
+          type: signal
+          signal: SIGTERM
+```
+
+The contents of the *notifications* section will allow us to do the requests to
+`/api/v1/namespaces/default/pods/signal-example/notify` with data like `{"name": "reconfigure"}` or
+`{"container": "my-container", "name": "reconfigure"}` or `{"name": "stop"}` etc.
```

Comment:
What is the container param in the POST?  Is it trying to disambiguate notifications of the same name?  I don't think it is needed, just assert that notification names must be unique across all containers in a pod.  Same as ports.